subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1634 @@
+
+gc_done:
+
+	/* restore the previous sector_cycle */
+	fs->sector_cycle = previous_cycle;
+
+	/* Make it possible to detect that gc has finished by writing a
+	 * gc done ate to the sector. In the field we might have zms systems
+	 * that do not have sufficient space to add this ate, so for these
+	 * situations avoid adding the gc done ate.
+	 */
+
+	if (fs->ate_wra >= (fs->data_wra + ate_size)) {
+		rc = zms_add_gc_done_ate(fs);
+		if (rc) {
+			return rc;
+		}
+	}
+
+	/* Erase the GC'ed sector when needed */
+	if (ebw_required) {
+		rc = zms_flash_erase_sector(fs, sec_addr);
+		if (rc) {
+			return rc;
+		}
+	}
+	/* Adding empty ate to the gc'ed sector increments fs->ate_wra
+	 * Restore the ate_wra of the current sector after this
+	 */
+	previous_ate_wra = fs->ate_wra;
+#ifdef CONFIG_ZMS_LOOKUP_CACHE
+	zms_lookup_cache_invalidate(fs, sec_addr >> ADDR_SECT_SHIFT);
+#endif
+	rc = zms_add_empty_ate(fs, sec_addr);
+	fs->ate_wra = previous_ate_wra;
+
+	return rc;
+}
+
+int zms_clear(struct zms_fs *fs)
+{
+	int rc;
+	uint64_t addr;
+	bool ebw_required =
+		flash_params_get_erase_cap(fs->flash_parameters) & FLASH_ERASE_C_EXPLICIT;
+
+	if (!fs->ready) {
+		LOG_ERR("zms not initialized");
+		return -EACCES;
+	}
+
+	for (uint32_t i = 0; i < fs->sector_count; i++) {
+		addr = (uint64_t)i << ADDR_SECT_SHIFT;
+		if (ebw_required) {
+			rc = zms_flash_erase_sector(fs, addr);
+			if (rc) {
+				return rc;
+			}
+		}
+		rc = zms_add_empty_ate(fs, addr);
+		if (rc) {
+			return rc;
+		}
+	}
+
+	/* zms needs to be reinitialized after clearing */
+	fs->ready = false;
+
+	return 0;
+}
+
+static int zms_init(struct zms_fs *fs)
+{
+	int rc;
+	struct zms_ate last_ate, first_ate, close_ate, empty_ate;
+	size_t ate_size;
+	/* Initialize addr to 0 for the case fs->sector_count == 0. This
+	 * should never happen as this is verified in zms_mount() but both
+	 * Coverity and GCC believe the contrary.
+	 */
+	uint64_t addr = 0U;
+	uint32_t i, closed_sectors = 0;
+	bool ebw_required =
+		flash_params_get_erase_cap(fs->flash_parameters) & FLASH_ERASE_C_EXPLICIT;
+
+	k_mutex_lock(&fs->zms_lock, K_FOREVER);
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+	/* step through the sectors to find a open sector following
+	 * a closed sector, this is where zms can write.
+	 */
+
+	for (i = 0; i < fs->sector_count; i++) {
+		addr = ((uint64_t)i << ADDR_SECT_SHIFT) + (uint32_t)(fs->sector_size - ate_size);
+
+		/* read the first and second ate to find out if sector is
+		 * open/close
+		 */
+		rc = zms_get_sector_header(fs, addr, &empty_ate, &close_ate);
+		if (rc) {
+			goto end;
+		}
+
+		/* update cycle count */
+		fs->sector_cycle = empty_ate.cycle_cnt;
+
+		if (zms_close_ate_valid(fs, &close_ate) && zms_empty_ate_valid(fs, &empty_ate) &&
+		    (close_ate.cycle_cnt == empty_ate.cycle_cnt)) {
+			/* closed sector */
+			closed_sectors++;
+			/* Let's check first that we support this ZMS version */
+			if (ZMS_GET_VERSION(empty_ate.metadata) != ZMS_DEFAULT_VERSION) {
```

Comment:
There is 32 bits of metadata available? Can we also put there "ZMS" aside from version? This way you can, in this line, do a basic check whether this is ZMS you are dealing with at all.

Reply:
@de-nordic you mean a magic number on one byte right ? I do not want to waist 32 bits on just "ZMS" I want to leave a space for more features and/or versions later